File: scripts/full-lean/probes/wasmtime-atomics.c
// Bridge prerequisite only. No shipping application uses this probe.
#include "wasmtime-memory-view.c"

static int atomic_address(probe_t *probe, uint64_t offset, uint32_t width, void **address) {
    if (width != 1 && width != 2 && width != 4 && width != 8) return 2;
    size_t size = wasmtime_sharedmemory_data_size(probe->memory);
    if (offset % width || offset > size || width > size - offset) return 2;
    *address = wasmtime_sharedmemory_data(probe->memory) + offset;
    return 0;
}

// Sequentially consistent integer operations on the SAME native backing as
// Wasm atomics. External ArrayBuffers are not SharedArrayBuffers; calling JS
// Atomics on those views is not a substitute for this native boundary.
// Operations: load, store, add, sub, and, or, xor, exchange, compareExchange.
// Bitwise RMW uses an explicit CAS loop: the system GCC 13.3 -O2 compiled the
// combined fetch-and/or/xor dispatch incorrectly in the preserved native-only
// control. The same control passed at -O0; all semantics remain tested at -O2.
int lasm_probe_atomic(probe_t *probe, uint64_t offset, uint32_t width, uint32_t operation,
                      uint64_t value, uint64_t comparison, uint64_t *result) {
    void *address;
    if (operation > 8 || atomic_address(probe, offset, width, &address)) return 2;
#define ACCESS(TYPE) do { \
    _Atomic(TYPE) *cell = address; \
    if (!atomic_is_lock_free(cell)) return 3; \
    TYPE before = 0; \
    switch (operation) { \
        case 0: before = atomic_load_explicit(cell, memory_order_seq_cst); break; \
        case 1: atomic_store_explicit(cell, (TYPE)value, memory_order_seq_cst); before = (TYPE)value; break; \
        case 2: before = atomic_fetch_add_explicit(cell, (TYPE)value, memory_order_seq_cst); break; \
        case 3: before = atomic_fetch_sub_explicit(cell, (TYPE)value, memory_order_seq_cst); break; \
        case 4: case 5: case 6: { \
            before = atomic_load_explicit(cell, memory_order_seq_cst); \
            TYPE desired; \
            do { desired = operation == 4 ? before & (TYPE)value \
                    : operation == 5 ? before | (TYPE)value : before ^ (TYPE)value; \
            } while (!atomic_compare_exchange_weak_explicit(cell, &before, desired, memory_order_seq_cst, memory_order_seq_cst)); \
            break; \
        } \
        case 7: before = atomic_exchange_explicit(cell, (TYPE)value, memory_order_seq_cst); break; \
        case 8: before = (TYPE)comparison; \
            atomic_compare_exchange_strong_explicit(cell, &before, (TYPE)value, memory_order_seq_cst, memory_order_seq_cst); break; \
    } \
    *result = before; return 0; \
} while (0)
    switch (width) { case 1: ACCESS(uint8_t); case 2: ACCESS(uint16_t);
        case 4: ACCESS(uint32_t); case 8: ACCESS(uint64_t); }
#undef ACCESS
    return 2;
}

probe_t *lasm_probe_atomics_new(char *error, size_t capacity) {
    probe_t *probe = lasm_probe_new(error, capacity);
    if (!probe) return NULL;
#define READ(BITS, SUFFIX) \
    " (func (export \"read" #BITS "\") (param i64 i64 i64) (result i64)" \
    " local.get 0 i64.atomic.load" SUFFIX ")"
    const char *wat =
        "(module (import \"env\" \"memory\" (memory i64 1 131072 shared))"
        READ(8, "8_u") READ(16, "16_u") READ(32, "32_u") READ(64, "")
        " (func (export \"add32\") (param $p i64) (param $n i64) (param i64) (result i64)"
        " (local $i i64) (block $done local.get $n i64.eqz br_if $done"
        " (loop $loop local.get $p i64.const 1 i64.atomic.rmw32.add_u drop"
        " local.get $i i64.const 1 i64.add local.tee $i local.get $n i64.lt_u br_if $loop))"
        " local.get $p i64.atomic.load32_u)"
        " (func (export \"add64\") (param $p i64) (param $n i64) (param i64) (result i64)"
        " (local $i i64) (block $done local.get $n i64.eqz br_if $done"
        " (loop $loop local.get $p i64.const 1 i64.atomic.rmw.add drop"
        " local.get $i i64.const 1 i64.add local.tee $i local.get $n i64.lt_u br_if $loop))"
        " local.get $p i64.atomic.load)"
        " (func (export \"wait32\") (param i64 i64 i64) (result i64)"
        " local.get 0 local.get 1 i32.wrap_i64 local.get 2 memory.atomic.wait32 i64.extend_i32_u)"
        " (func (export \"wait64\") (param i64 i64 i64) (result i64)"
        " local.get 0 local.get 1 local.get 2 memory.atomic.wait64 i64.extend_i32_u)"
        " (func (export \"notify\") (param i64 i64 i64) (result i64)"
        " local.get 0 local.get 1 i32.wrap_i64 memory.atomic.notify i64.extend_i32_u))";
#undef READ
    wasm_byte_vec_t bytes;
    if (error_text(wasmtime_wat2wasm(wat, strlen(wat), &bytes), NULL, error, capacity)) goto fail;
    wasmtime_module_t *module = NULL;
    wasmtime_error_t *compiled = wasmtime_module_new(probe->engine, (uint8_t *)bytes.data, bytes.size, &module);
    wasm_byte_vec_delete(&bytes);
    if (error_text(compiled, NULL, error, capacity)) goto fail;
    wasmtime_module_delete(probe->module); probe->module = module;
    return probe;
fail:
    lasm_probe_delete(probe); return NULL;
}

static int atomic_wasm(probe_t *probe, const char *name, uint64_t offset, uint64_t value,
                       int64_t timeout, uint64_t *result, char *error, size_t capacity, bool bounded) {
    uint32_t width = strstr(name, "64") ? 8 : strstr(name, "16") ? 2 : strstr(name, "read8") ? 1 : 4;
    void *address;
    if (atomic_address(probe, offset, width, &address)) return 2;
    if (strncmp(name, "add", 3) == 0 && value > 200000) return 2;
    if (bounded && strncmp(name, "wait", 4) == 0 && (timeout < 0 || timeout > INT64_C(5000000000))) return 2;
    wasmtime_store_t *store = wasmtime_store_new(probe->engine, NULL, NULL);
    if (!store) { snprintf(error, capacity, "store allocation failed"); return 1; }
    wasmtime_context_t *context = wasmtime_store_context(store);
    wasmtime_sharedmemory_t *memory = wasmtime_sharedmemory_clone(probe->memory);
    wasmtime_extern_t imported = { .kind = WASMTIME_EXTERN_SHAREDMEMORY, .of.sharedmemory = memory };
    wasmtime_instance_t instance;
    wasm_trap_t *trap = NULL;
    wasmtime_error_t *instance_error = wasmtime_instance_new(context, probe->module, &imported, 1, &instance, &trap);
    int status = error_text(instance_error, trap, error, capacity);
    if (!status) {
        wasmtime_extern_t item;
        if (!wasmtime_instance_export_get(context, &instance, name, strlen(name), &item)) {
            snprintf(error, capacity, "unknown atomic operation"); status = 2;
        } else {
            if (item.kind != WASMTIME_EXTERN_FUNC) {
                snprintf(error, capacity, "atomic export is not a function");
                wasmtime_extern_delete(&item); status = 2; goto cleanup;
            }
            wasmtime_val_t inputs[3] = {
                { .kind = WASMTIME_I64, .of.i64 = (int64_t)offset },
                { .kind = WASMTIME_I64, .of.i64 = (int64_t)value },
                { .kind = WASMTIME_I64, .of.i64 = timeout },
            }, returned;
            trap = NULL;
            wasmtime_error_t *called = wasmtime_func_call(context, &item.of.func, inputs, 3, &returned, 1, &trap);
            status = error_text(called, trap, error, capacity);
            if (!status) {
                if (returned.kind != WASMTIME_I64) {
                    snprintf(error, capacity, "atomic result is not i64"); status = 2;
                } else *result = (uint64_t)returned.of.i64;
                wasmtime_val_unroot(&returned);
            }
            wasmtime_extern_delete(&item);
        }
    }
cleanup:
    wasmtime_sharedmemory_delete(memory); wasmtime_store_delete(store);
    return status;
}

// Preserve the original bounded probe contract and its invalid-input controls.
int lasm_probe_atomic_wasm(probe_t *probe, const char *name, uint64_t offset, uint64_t value,
                           int64_t timeout, uint64_t *result, char *error, size_t capacity) {
    return atomic_wasm(probe, name, offset, value, timeout, result, error, capacity, true);
}

// Blocking adapter boundary. -1 is Wasm's infinite timeout; all exercised equal
// waits remain bounded by a finite deadline. Invalid-address checks still apply.
int lasm_probe_atomic_wait(probe_t *probe, uint64_t offset, uint32_t width, uint64_t value,
                           int64_t timeout, uint64_t *result, char *error, size_t capacity) {
    if ((width != 4 && width != 8) || timeout < -1) return 2;
    return atomic_wasm(probe, width == 8 ? "wait64" : "wait32", offset, value,
                       timeout, result, error, capacity, false);
}

File: scripts/full-lean/probes/wasmtime-atomics.mjs
// Native atomic access to sparse memory64 backing; not a shipping backend.
import assert from 'node:assert/strict';
import { createRequire } from 'node:module';
import { Worker, isMainThread, parentPort, workerData } from 'node:worker_threads';
import { fileURLToPath } from 'node:url';

const ffi = createRequire(import.meta.url)('koffi');
const libraryPath = isMainThread ? process.argv[2] : workerData.libraryPath;
const library = ffi.load(libraryPath);
const create = library.func('void *lasm_probe_atomics_new(char *error, size_t capacity)');
const destroy = library.func('void lasm_probe_delete(void *probe)');
const access = library.func('int lasm_probe_atomic(void *probe, uint64_t offset, uint32_t width, uint32_t operation, uint64_t value, uint64_t comparison, _Out_ uint64_t *result)');
const invoke = library.func('int lasm_probe_atomic_wasm(void *probe, str name, uint64_t offset, uint64_t value, int64_t timeout, _Out_ uint64_t *result, char *error, size_t capacity)');
const error = Buffer.alloc(4096), message = () => error.toString('utf8').split('\0')[0];
function atomic(probe, offset, width, operation, value = 0n, comparison = 0n) {
  const result = [0]; assert.equal(access(probe, offset, width, operation, value, comparison, result), 0);
  return BigInt(result[0]);
}
function wasm(probe, name, offset, value = 0n, timeout = 0n) {
  const result = [0]; assert.equal(invoke(probe, name, offset, value, timeout, result, error, error.length), 0, message());
  return BigInt(result[0]);
}

if (!isMainThread) {
  const { probe, mode, width, offset, iterations } = workerData;
  parentPort.once('message', () => {
    let result;
    if (mode === 'wait') result = wasm(probe, 'wait' + width * 8, offset, 0n, 2_000_000_000n);
    else if (mode === 'wasm-add') result = wasm(probe, 'add' + width * 8, offset, BigInt(iterations));
    else {
      assert.equal(mode, 'host-add');
      for (let i = 0; i < iterations; i++) atomic(probe, offset, width, 2, 1n);
      result = atomic(probe, offset, width, 0);
    }
    parentPort.postMessage({ result: result.toString() });
  });
  parentPort.postMessage({ ready: true });
} else {
  const probe = create(error, error.length); assert.ok(probe, message());
  const live = new Set(), checks = [];
  function worker(task) {
    const child = new Worker(fileURLToPath(import.meta.url), {
      workerData: { libraryPath, probe: ffi.address(probe), ...task },
    });
    live.add(child);
    let ready, failReady, finish, fail, received;
    const initialized = new Promise((resolve, reject) => { ready = resolve; failReady = reject; });
    const completed = new Promise((resolve, reject) => { finish = resolve; fail = reject; });
    completed.catch(() => {});
    const timer = setTimeout(() => { failReady(new Error('Atomic worker deadline')); fail(new Error('Atomic worker deadline')); }, 10_000);
    child.on('message', value => { if (value.ready) ready(); else received = value; });
    child.once('error', err => { failReady(err); fail(err); });
    child.once('exit', code => {
      live.delete(child); clearTimeout(timer);
      if (code !== 0 || !received) { const err = new Error('Atomic worker exited without result: ' + code); failReady(err); fail(err); }
      else finish(received);
    });
    return { initialized, completed, go: () => child.postMessage({ go: true }) };
  }
  try {
    for (const base of [0, 2 ** 32]) {
      for (const width of [1, 2, 4, 8]) {
        const offset = base + width * 16;
        const Type = ({ 1: Uint8Array, 2: Uint16Array, 4: Uint32Array, 8: BigUint64Array })[width];
        const reference = new Type(new SharedArrayBuffer(width));
        const js = value => width === 8 ? value : Number(value);
        const maximum = (1n << BigInt(width * 8)) - 1n;
        const sequence = [['store', 1, maximum], ['add', 2, 1n], ['sub', 3, 1n],
          ['store', 1, 251n], ['add', 2, 253n], ['sub', 3, 507n],
          ['and', 4, 0x1234n], ['or', 5, 0x80n], ['xor', 6, 0x5an], ['exchange', 7, 44n],
          ['compareExchange', 8, 55n, 44n], ['compareExchange', 8, 66n, 44n]];
        for (const [name, op, value, comparison = 0n] of sequence) {
          const before = name === 'compareExchange' ? Atomics.compareExchange(reference, 0, js(comparison), js(value))
            : Atomics[name](reference, 0, js(value));
          const actual = atomic(probe, offset, width, op, value, comparison);
          // The primitive store reports stored bits; the future JS facade must
          // implement Atomics.store's distinct converted-input return value.
          const context = JSON.stringify({ offset, width, name, value: String(value), comparison: String(comparison) });
          if (name !== 'store') assert.equal(actual, BigInt(before), context + ': returned value');
          assert.equal(atomic(probe, offset, width, 0), BigInt(Atomics.load(reference, 0)), context + ': stored value');
          assert.equal(wasm(probe, 'read' + width * 8, offset), BigInt(Atomics.load(reference, 0)), context + ': Wasm read');
        }
        checks.push({ kind: 'scalar-native-js-and-wasm', offset, width, operations: sequence.length });
      }
      for (const width of [4, 8]) {
        const offset = base + 256 + width * 16;
        atomic(probe, offset, width, 1, 0n);
        assert.equal(wasm(probe, 'wait' + width * 8, offset, 1n, 0n), 1n);
        assert.equal(wasm(probe, 'wait' + width * 8, offset, 0n, 1_000_000n), 2n);
        const pending = worker({ mode: 'wait', width, offset }); await pending.initialized; pending.go();
        const deadline = Date.now() + 1500;
        let notified = 0n;
        while (!notified && Date.now() < deadline) {
          notified = wasm(probe, 'notify', offset, 1n);
          if (!notified) await new Promise(resolve => setTimeout(resolve, 2));
        }
        assert.equal(notified, 1n); assert.deepEqual(await pending.completed, { result: '0' });
        checks.push({ kind: 'wait-notify', offset, width, notEqual: 1, timeout: 2, awakened: 0 });
        atomic(probe, offset, width, 1, 0n);
        const first = worker({ mode: 'host-add', offset, width, iterations: 10000 });
        const second = worker({ mode: 'wasm-add', offset, width, iterations: 10000 });
        await Promise.all([first.initialized, second.initialized]); first.go(); second.go();
        await Promise.all([first.completed, second.completed]);
        assert.equal(atomic(probe, offset, width, 0), 20000n);
        assert.equal(wasm(probe, 'read' + width * 8, offset), 20000n);
        checks.push({ kind: 'two-worker-native-and-wasm-add', offset, width, increments: 20000 });
      }
    }
    for (const [offset, width, operation] of [[1n, 8, 0], [2n ** 64n - 1n, 8, 0],
      [BigInt(2 ** 32 + 2 * 65536), 4, 0], [0n, 3, 0], [0n, 4, 9]])
      assert.equal(access(probe, offset, width, operation, 0, 0, [0]), 2);
    for (const [name, offset, value, timeout] of [
      ['read64', 1n, 0n, 0n], ['read64', 2n ** 64n - 1n, 0n, 0n],
      ['unknown', 0n, 0n, 0n], ['wait64', 0n, 0n, 5_000_000_001n],
      ['add64', 0n, 200001n, 0n],
    ]) assert.equal(invoke(probe, name, offset, value, timeout, [0], error, error.length), 2);
    console.log(JSON.stringify({ scope: 'Sparse memory64 host atomic primitives and Wasm wait/notify; no full JS Atomics facade or Lean application integration',
      engine: process.versions.bun ? 'bun' : process.versions.deno ? 'deno' : 'node',
      version: process.versions.bun ?? process.versions.deno ?? process.versions.node,
      accessibleGuestBytes: 2 ** 32 + 2 * 65536, touchedCells: 12, checks,
      invalidNativeAddressesAndOperationsRejected: 5, invalidWasmInputsRejected: 5 }));
  } finally {
    await Promise.all([...live].map(child => child.terminate()));
    destroy(probe);
  }
}
